### DISCUSSION TITLE
Setup with a generic rack application?

### DISCUSSION THREAD

**Discussion by burlesona** [ElMassimo/vite_ruby#121] [Q&A] [answered]
I'm specifically interested in using Vite Ruby with Roda. I followed the guide, but there's very little content for using `vite_ruby` alone. I get the idea that what it's doing is setting up Vite to serve only assets (not a root html file) and that we can load those assets on localhost:3036/vite-dev.

What is not clear is how this should be integrated into a generic rack app. I'm trying to read the source code and connect the dots, but I'm not quite there.

Presumably I need to use the ViteRuby's asset methods to link to `localhost:3036/vite-dev/application.js` while in development, then in production use the manifest to turn that same link into `/application-<somehash>.js`.

I see how in the framework plugins there are template helpers that do this, and I understand why in a plain Rack app there's no way to automatically provide such helpers, but I was expecting to find something like `Vite.asset_path(...)` that I could use, and I'm not seeing that. Am I just missing it?

Also it seems like the browser will not like doing that with a different host in development, but I see in the source that there's a proxy dev server which I assume is meant to be used as middleware to proxy requests from localhost:<rack server port> to localhost:<vite dev port> so that the browser doesn't know there are two things running. So, step 1 for a Rack app is `use ViteRuby::DevServerProxy`?

I think I'm close here, but would love a little help if anyone can fill in some of these blanks for me. Thank you!

**Comment by ElMassimo** [comment, marked as answer]
Hi Andrew!

By default, running `bundle exec vite install` in a plain Rack app will attempt to [configure the dev server proxy](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/cli/install.rb#L39-L44).

Once the proxy is in place, the next step is to use [`ViteRuby.instance.manifest`](https://github.com/ElMassimo/vite_ruby/blob/main/vite_rails/lib/vite_rails/tag_helpers.rb#L54-L57) and use its public methods to obtain the URLs corresponding to the resources, and render `script` and `style` tags.

For example, for stylesheets the most basic implementation would be:

```erb
<link rel="stylesheet" href="<%= vite_manifest.path_for('styles', type: :stylesheet) %>" media="all"/>
```

You would typically encapsulate this as:

```erb
<%= vite_stylesheet_tag 'styles' %>
```

For javascript it's a bit more complex, as `resolve_entries` will also return CSS files that should be injected as well, but [it can be done](https://github.com/ElMassimo/jekyll-vite/blob/main/lib/jekyll/vite/tags.rb#L132-L147).

<hr/>

I initially intended to create `vite_roda`, but the lack of conventions make it hard to provide a seamless integration.

If you find some primitives in Roda that would allow to implement this [concisely](https://github.com/ElMassimo/vite_ruby/blob/main/vite_padrino/lib/vite_padrino/tag_helpers.rb), let me know and I'll reconsider it.

**Comment by Pistos** [comment]
Recently created (not mine): https://github.com/holamendi/vite-roda . It's working for me for simple `vite_javascript_tag`, at least.
